src/Front/Ast.hs

```diff
@@ -108,88 +110,69 @@ data Literal
 
 -- | Annotate 'Bind' with additional types
 annBind :: Bind -> [Typ] -> Bind
-annBind (Bind p ts) ts' = Bind p (ts'++ts)
+annBind (Bind p ts) ts' = Bind p (ts' ++ ts)
 
 -- | Collect a curried application into the function and its list of arguments.
 collectTApp :: Typ -> (Typ, [Typ])
 collectTApp (TApp lhs rhs) = (lf, la ++ [rhs])
   where (lf, la) = collectTApp lhs
 collectTApp t = (t, [])
 
-instance Pretty Program where
-  pretty (Program _) = undefined
+instance Show Program where
+  show (Program defs) = concatMap (\df -> show (pretty df) ++ "\n\n") defs
 
 instance Pretty Definition where
-  pretty = undefined
-    {-
-  pretty (Function id formals body r) =
-    let ret =
-          (case r of
-            ReturnType  t -> pretty "->" <+> pretty t
-            CurriedType t -> pretty ":" <+> pretty t
-          )
-    in  nest
-          2
-          (vsep
-            [ pretty id <> tupled (map pretty formals) <+> ret <+> pretty '='
-            , pretty body
-            ]
-          )
-          -}
+  pretty (DefFn id formals r body) = nest
+    2
+    (vsep
+      [ pretty id <> tupled (map pretty formals) <+> pretty r <+> pretty '='
+      , pretty body
+      ]
+    )
+  pretty (DefPat b body) = pretty b <+> pretty "=" <+> pretty body
 
 instance Pretty Bind where
-  pretty = undefined
-  {-
-  pretty (Bind id mty) =
-    let prettyId = pretty id
-    in  case mty of
-          Just ty -> prettyId <+> pretty ':' <+> pretty ty
-          Nothing -> prettyId
-  pretty (TupBind binds mty) =
-    let prettyTup = hsep (punctuate comma $ map pretty binds)
-    in  case mty of
-          Just ty -> parens prettyTup <+> pretty ':' <+> pretty ty
-          Nothing -> prettyTup
-  -}
-
-  {-
-  pretty (PId      s) = pretty s
-  pretty (PLiteral l) = pretty l
-  pretty PWildcard    = pretty '_'
-  pretty (PAs  v p )  = pretty v <> pretty '@' <> pretty p
-  pretty (PCon c ps)  = pretty c <+> hsep (map pretty ps)
-  -}
+  pretty (Bind pat []) = pretty pat -- TODO(hans) why is ty a list?
+  pretty (Bind pat [ty]) = pretty pat <+> pretty ":" <+> pretty ty
+  pretty (Bind pat tys) =
+    pretty pat
+      <+> pretty ": ["
+      <+> hsep (punctuate comma $ map pretty tys)
+      <+> pretty "]"
+
+instance Pretty BindPat where
+  pretty BindWildcard   = pretty '_'
+  pretty (BindId  s   ) = pretty s
+  pretty (BindLit l   ) = pretty l
+  pretty (BindAs b p  ) = pretty b <> pretty '@' <> pretty p -- TODO(hans): Parens around p?
```

Comment:
Yeah you may as well defensively add parens around everything.

We will eventually take another pass that adds precedence to things.